allow back button customization. more than just title change and tint.

pvinis opened the pull request:
This PR allows the user to use any react element as a back button.
We can already use something like
```
const Navigator = StackNavigator(
  {
    Splash: { screen: SplashScreen },
    Login: { screen: LoginScreen },
    Main: { screen: MainScreen },
  },
  {
    initialRouteName: 'Splash',
    navigationOptions: {
      headerMode: 'screen',
      headerStyle: {
        backgroundColor: Colors.navBarBackground,
        borderBottomWidth: 1 / PixelRatio.get(),
        borderBottomColor: Colors.separator,
      },
      headerTitleStyle: {
        color: Colors.textBlue,
        fontFamily: Fonts.type.base,
      },
```
Now we can also add something like
```
    headerBack: <Image source={Images.iconBack} />,
```
or like
```
    headerBack: (
      <View style={{
        alignItems: 'center',
        flexDirection: 'row',
      }}>
        <Image source={Images.iconBack} />
        <Text style={{paddingRight: 10, }} numberOfLines={1}>BACK</Text>
      </View>
    ),
```
This will be wrapped and made into a button that goes back.
The difference with `headerLeft` is basically that `headerLeft` can be anything, and added in any screen. `headerBack` is always a button that goes back, that contains any custom element the user gives, and it will only appear on non-zero index screens.

Without a custom `headerBack` (regular back button, automatically rendered, as before):
![img_0052](https://cloud.githubusercontent.com/assets/100233/26007178/b1c8ee1e-373f-11e7-8f28-087338452b00.PNG)
(`PavlosTv` is the title of the previous screen)

With a custom `headerBack` (custom element, wrapped as a `goBack` button):
![img_0051](https://cloud.githubusercontent.com/assets/100233/26007123/8246ae42-373f-11e7-84de-cf0c71ec84a0.PNG)

With a custom `headerBack` (custom element, wrapped as a `goBack` button):
![img_0053](https://cloud.githubusercontent.com/assets/100233/26007154/a22ea2f0-373f-11e7-88e3-fd7cdc73f46d.PNG)

koenpunt commented:
Isn't the same achievable with something like:

```js
navigationOptions: ({ navigation }) => {
  const left = navigation.state.index === 0 ? null : <MyBackButton/>;
  return { headerLeft: left };
}
```

matthamil commented:
This feature already exists for react-navigation, as @koenpunt has mentioned. If this is a new feature, how is this different from the existing API as seen here https://github.com/react-community/react-navigation/blob/ff99d6ec0a2d7a62adfd8d1e6a3c906fcedf294a/src/views/Header.js#L151

https://reactnavigation.org/docs/navigators/stack#headerLeft

koenpunt commented:
There might be a caveat here, because I think the index isn't passed to the `navigationOptions`. To workaround that fact you could set `headerLeft: null` on the root navigator to explicitly disable the back button.

pvinis commented:
I will check, but I also think the index is not in there. And it feels cleaner this way anyway. we have `headerLeft` for `put this on the left always`, which could be hamburger button, or search or something, and we now have `headerBack`, which is basically customizing the regular `back` button. Dont you guys agree?

aksonov commented:
@koenpunt Is it bug that index is not passed to navigationOptions? Is it going to be fixed?

pvinis commented:
god.. this repo needs more active people that work on issues and prs.

matthamil commented on src/views/Header.js:
```diff
@@ -119,25 +120,35 @@ class Header extends React.PureComponent<void, HeaderProps, HeaderState> {
     const width = this.state.widths[props.scene.key]
       ? (this.props.layout.initWidth - this.state.widths[props.scene.key]) / 2
       : undefined;
+    var backButtonElement = options.headerBack;
```
I think it would be better to extract the logic that determines `backButtonElement` into a function that returns the value of what `backButtonElement` should be rather than imperatively determining what its value is. See `this._getBackButtonTitleString` as an example.

pvinis commented:
should be better now

nonameolsson commented:
This is a great PR! Could it be fixed and merged?

pvinis commented:
i will try to fix it in the weekend, even though im planning to move away from react-navigation soon.

pvinis commented:
closing in favor of #2869.
i fixed everything, added a test/example, and it sits on top of master, which is one commit ahead of beta.16, so it should be pretty easy to merge.
thanks again.
